rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2127 @@
+        // Update the stub params based on the packaet type
+        match self.get_hdr_type() {
+            Some(x) => match x {
+                DCERPC_TYPE_REQUEST => {
+                    if let Some(ref mut req) = self.request {
+                        retval = evaluate_stub_params(
+                            input,
+                            input_len,
+                            hdrpfcflags,
+                            padleft,
+                            &mut req.stub_data_buffer,
+                            &mut req.stub_data_buffer_len,
+                            &mut req.stub_data_buffer_reset,
+                        );
+                    }
+                }
+                DCERPC_TYPE_RESPONSE => {
+                    if let Some(ref mut resp) = self.response {
+                        retval = evaluate_stub_params(
+                            input,
+                            input_len,
+                            hdrpfcflags,
+                            padleft,
+                            &mut resp.stub_data_buffer,
+                            &mut resp.stub_data_buffer_len,
+                            &mut resp.stub_data_buffer_reset,
+                        );
+                    }
+                }
+                _ => {
+                    SCLogDebug!("Unrecognized packet type");
+                    return 0;
+                }
+            },
+            None => {
+                return 0;
+            }
+        }
+        // Update the remaining fragment length
+        self.padleft -= retval;
+
+        retval
+    }
+
+    /// Handles stub data for both request and response.
+    ///
+    /// Arguments:
+    /// * `input`:
+    ///           type: u8 vector slice.
+    ///    description: bytes left *after* parsing header.
+    /// * `bytes_consumed`:
+    ///           type: 32 bit integer.
+    ///    description: bytes consumed *after* parsing header.
+    /// * `dir`:
+    ///           type: 8 bit unsigned integer.
+    ///    description: direction whose stub is supposed to be handled.
+    ///
+    /// Return value:
+    /// * Success: Number of bytes successfully parsed.
+    /// * Failure: -1 in case fragment length defined by header mismatches the data.
+    pub fn handle_common_stub(&mut self, input: &[u8], bytes_consumed: i32, dir: u8) -> i32 {
+        let fraglen = self.get_hdr_fraglen().unwrap_or(0);
+        if fraglen < bytes_consumed as u16 + DCERPC_HDR_LEN {
+            return -1;
+        }
+        self.padleft = fraglen - DCERPC_HDR_LEN - bytes_consumed as u16; // bytes processed so far = header length
+        let mut input_left = input.len() as i32 - bytes_consumed;
+        let mut parsed: i32 = bytes_consumed; // parsed bytes
+        while input_left > 0 && parsed < fraglen as i32 {
+            let retval = self.handle_stub_data(&input[parsed as usize..], input_left as u16);
+            if retval > 0 || retval > input_left as u16 {
```

Comment:
same as with udp, should this be `retval >= input_left`?

Reply:
Wait a minute: if retval > input_left, input_left -= retval would underflow input_left, right?

Reply:
input_left I think happened because of replacing a lot of occurences while renaming things but it would still be incorrect. I think neither of the cases reached this part so nothing failed. I think it should be `retval > 0 && retval < input_len`